core/distributed/vector.cpp

```diff
@@ -0,0 +1,392 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/vector.hpp>
+
+
+#include "core/distributed/vector_kernels.hpp"
+#include "core/matrix/dense_kernels.hpp"
+
+
+namespace gko {
+namespace distributed {
+namespace vector {
+
+
+GKO_REGISTER_OPERATION(compute_norm2_sqr, dense::compute_norm2_sqr);
+GKO_REGISTER_OPERATION(compute_sqrt, dense::compute_sqrt);
+GKO_REGISTER_OPERATION(outplace_absolute_dense, dense::outplace_absolute_dense);
+GKO_REGISTER_OPERATION(build_local, distributed_vector::build_local);
+
+
+}  // namespace vector
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Vector<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
+    const LinOp* b, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Vector<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
+    const LinOp* alpha, const LinOp* b, const LinOp* beta, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Vector<ValueType, LocalIndexType, GlobalIndexType>::Vector(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    std::shared_ptr<const Partition<LocalIndexType, GlobalIndexType>> partition,
+    dim<2> global_size, dim<2> local_size, size_type stride)
+    : EnableLinOp<
+          Vector<ValueType, LocalIndexType, GlobalIndexType>>{exec,
+                                                              global_size},
+      DistributedBase{comm},
+      partition_{
+          partition
+              ? std::move(partition)
+              : gko::share(
+                    Partition<LocalIndexType, GlobalIndexType>::create(exec))},
+      local_{exec, local_size,
+             stride != invalid_index<size_type>() ? stride : local_size[1]}
+{}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType,
+          typename LocalMtxType>
+void read_local_impl(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    const Partition<LocalIndexType, GlobalIndexType>* partition,
+    const size_type num_cols,
+    const Array<matrix_data_entry<ValueType, GlobalIndexType>>& global_data,
+    LocalMtxType* local_mtx)
+{
+    auto rank = comm.rank();
+
+    Array<matrix_data_entry<ValueType, LocalIndexType>> local_data{exec};
+    exec->run(vector::make_build_local(global_data, partition, rank, local_data,
+                                       ValueType{}));
+
+    auto local_rows = static_cast<size_type>(partition->get_part_size(rank));
+    dim<2> local_size{local_rows, num_cols};
+    local_mtx->read({local_size, local_data});
```

Comment:
I think that should create a `device_matrix_data`, so that should be the same as writing directly to dense. I should use `move` to remove the unnecessary copy of the array though.

Reply:
using device_matrix_data is at least a 2x overhead for double/int, and the complexity of the algorithm is really not necessary, because we can skip storing anything in a variable-size array

Reply:
Ok, I think I get it now, you mean passing `local_mtx->get_values()` into `build_local` right?

Reply:
or passing local_mtx directly, yes